sesman/session.c

```diff
@@ -598,28 +598,48 @@ session_start_fork(tbus data, tui8 type, struct SCP_SESSION *s)
                 if (s->directory != 0)
                 {
-                    if (s->directory[0] != 0)
+                    if (g_strncasecmp("NO", g_cfg->sec.allowed_alternate_shells, 3) != 0)
                     {
-                        g_set_current_dir(s->directory);
+                        if (s->directory[0] != 0)
+                        {
+                            g_set_current_dir(s->directory);
+                        }
+                    }
+                    else
+                    {
+                        LOG(LOG_LEVEL_WARNING,
+                            "Directory change to %s requested, but not "
+                            "allowed by AllowedAlternateShells config value.",
+                            s->directory);
                     }
                 }
                 if (s->program != 0 && s->program[0] != 0)
                 {
-                    if (g_strchr(s->program, ' ') != 0 || g_strchr(s->program, '\t') != 0)
+                    if (g_strncasecmp("NO", g_cfg->sec.allowed_alternate_shells, 3) != 0)
                     {
-                        LOG(LOG_LEVEL_INFO,
-                            "Starting user requested window manager on "
-                            "display %d with embedded arguments using a shell: %s",
-                            display, s->program);
-                        const char *params[] = {"sh", "-c", s->program, NULL};
-                        g_execvp("/bin/sh", (char **)params);
+                        if (g_strchr(s->program, ' ') != 0 || g_strchr(s->program, '\t') != 0)
+                        {
+                            LOG(LOG_LEVEL_INFO,
+                                "Starting user requested window manager on "
+                                "display %d with embedded arguments using a shell: %s",
+                                display, s->program);
+                            const char *params[] = {"sh", "-c", s->program, NULL};
+                            g_execvp("/bin/sh", (char **)params);
+                        }
+                        else
+                        {
+                            LOG(LOG_LEVEL_INFO,
+                                "Starting user requested window manager on "
+                                "display %d: %s", display, s->program);
+                            g_execlp3(s->program, s->program, 0);
+                        }
                     }
                     else
                     {
-                        LOG(LOG_LEVEL_INFO,
-                            "Starting user requested window manager on "
-                            "display %d: %s", display, s->program);
-                        g_execlp3(s->program, s->program, 0);
+                        LOG(LOG_LEVEL_WARNING,
+                            "Shell %s requested by user, but not allowed by "
+                            "AllowAlternateShells config value.",
+                            s->program);
                     }
```

Comment:
I don't think anything is being started if this happens. By all means log the warning, but then please take the default action.

Reply:
It just falls through to the next `if (g_cfg->enable_user_wm)`, and if that isn't supplied the default window manager is started.

I have now updated to debian unstable, so I checked that this is indeed the case for 0.9.21.1 (plus this patch).

Reply:
Oh fine - of course it does. Thanks.